src/main/java/org/opensearch/security/privileges/PitAccessEvaluator.java

```diff
@@ -0,0 +1,344 @@
+                            ImmutableSet.of(), SearchRequest.DEFAULT_INDICES_OPTIONS);
+
+            final Set<String> allPermittedIndices = securityRoles.reduce(pitResolved,
+                    user, new String[]{action}, resolver, clusterService);
+            if(isDebugEnabled) {
+                log.debug("Evaluating PIT ID : " + pitId );
+            }
+            if (allPermittedIndices.size() == INDICES_SET.size()) {
+                if(isDebugEnabled) {
+                    log.debug(" Permitting PIT ID : " + pitId);
+                }
+                permittedPits.add(pitToPitInfoMap.get(pitId));
+            }
+        }
+        if (permittedPits.size() > 0) {
+            ((GetAllPitNodesRequest) request).setGetAllPitNodesResponse(new GetAllPitNodesResponse(permittedPits,
+                    ((GetAllPitNodesRequest) request).getGetAllPitNodesResponse()));
+            presponse.allowed = true;
+            presponse.markComplete();
+        }
+        return presponse;
+    }
+
+    /**
+     * Handle access for 'delete all PITs' operation
+     */
+    private PrivilegesEvaluatorResponse handleDeleteAllPitAccess(DeletePitRequest deletePitRequest, ClusterService clusterService,
+                                          User user, SecurityRoles securityRoles, final String action,
+                                          IndexNameExpressionResolver resolver,
+                                          PrivilegesEvaluatorResponse presponse) throws InterruptedException {
+        List<String> permittedPits = new ArrayList<>();
+        List<String> pitIds = getAllPitIds();
+        // allow delete pit operation if there are no pits in the cluster ( response should be empty )
+        if(pitIds.isEmpty()) {
+            deletePitRequest.clearAndSetPitIds(pitIds);
+            presponse.allowed = true;
+            presponse.markComplete();
+        }
+        Map<String, String[]> pitToIndicesMap = OpenSearchSecurityPlugin.GuiceHolder.getPitService().getIndicesForPits(pitIds);
+        for (String pitId : pitIds) {
+            String[] indices = pitToIndicesMap.get(pitId);
+            HashSet<String> indicesSet = new HashSet<>(Arrays.asList(indices));
+            Set<String> allPermittedIndices = getPermittedIndices(indicesSet, clusterService, user,
+                    securityRoles, action, resolver);
+            // user should have permissions for all indices associated with PIT, only then add PIT ID as permitted PIT
+            if(isDebugEnabled) {
+                log.debug("Evaluating PIT : " + pitId );
+            }
+            if (allPermittedIndices.size() == indicesSet.size()) {
+                if(isDebugEnabled) {
+                    log.debug(" Permitting PIT : " + pitId);
+                }
+                permittedPits.add(pitId);
+            }
+        }
+        // If there are any PITs for which the user has access to, then allow operation otherwise fail.
+        if(permittedPits.size() > 0) {
```

Comment:
When a user has right permissions, even if permittedPits.size() equals to 0, the request also should be authorized and get empty in the response.

Reply:
@cliu123 i tried making this work, but looks like for index actions, there is no way to authorize the user when there are no indices to validate against. 

dnfOfEnabled flag is used for indices actions where for empty we permit , otherwise we throw forbidden error.

can you please suggest if there are any alternate approaches ?

Reply:
Here is a potential alternative approach:
1) Collect authorization information for the user
2) Does the user have access to any PITs? (If no- permission is denied)
3) Does the user have access to the PITs in the request? (If no- permission is denied)

Reply:
That is what we are doing currently when the config is false ( which is default ).
This config is used in search and other index actions as well, so can we please take a look at this config? it is designed not to throw forbidden in case of empty results.